File: promotion/user-service/src/main/java/com/study/userservice/controller/AuthController.java
package com.study.userservice.controller;

import com.study.userservice.dto.UserDto;
import com.study.userservice.entity.User;
import com.study.userservice.service.JWTService;
import com.study.userservice.service.UserService;
import io.jsonwebtoken.Claims;
import java.util.Collections;
import java.util.Map;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/users")
public class AuthController {

  private final JWTService jwtService;
  private final UserService userService;

  public AuthController(JWTService jwtService, UserService userService) {
    this.jwtService = jwtService;
    this.userService = userService;
  }

  @PostMapping("/login")
  public ResponseEntity<?> login(
      @RequestBody UserDto.LoginRequest request) {
    User user = userService.authenticate(request.getEmail(), request.getPassword());
    String token = jwtService.generateToken(user);
    return ResponseEntity.ok(UserDto.LoginResponse.builder()
        .token(token)
        .user(UserDto.Response.from(user))
        .build());
  }

  @PostMapping("/validate-token")
  public ResponseEntity<?> validateToken(
      @RequestBody UserDto.TokenRequest request) {
    Claims claims = jwtService.validateToken(request.getToken());
    return ResponseEntity.ok(UserDto.TokenResponse.builder()
        .email(claims.getSubject())
        .valid(true)
        .role(claims.get("role", String.class))
        .build());
  }

  @PostMapping("/refresh-token")
  public ResponseEntity<Map<String, String>> refreshToken(
      @RequestBody UserDto.TokenRequest tokenRequest) {
    String newToken = jwtService.refreshToken(tokenRequest.getToken());
    return ResponseEntity.ok(Collections.singletonMap("token", newToken));
  }
}

File: promotion-kotlin/user-service-kotlin/src/main/kotlin/com/study/userservicekotlin/controller/AuthController.kt
package com.study.userservicekotlin.controller

import com.study.userservicekotlin.controller.dto.LoginRequest
import com.study.userservicekotlin.controller.dto.LoginResponse
import com.study.userservicekotlin.controller.dto.TokenRequest
import com.study.userservicekotlin.controller.dto.TokenResponse
import com.study.userservicekotlin.controller.dto.toResponse
import com.study.userservicekotlin.service.JwtService
import com.study.userservicekotlin.service.UserService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v2/users")
class AuthController(
    private val jwtService: JwtService,
    private val userService: UserService,
) {

    @PostMapping("/login")
    fun login (
        @RequestBody loginRequest: LoginRequest,
    ): ResponseEntity<*> {
        val user = userService.authenticate(loginRequest.email, loginRequest.password)
        val generateToken = jwtService.generateToken(user)
        return ResponseEntity.ok(LoginResponse(generateToken, user.toResponse()))
    }

    @PostMapping("/validate-token")
    fun validateToken(
        @RequestBody request: TokenRequest,
    ): ResponseEntity<*> {
        val claims = jwtService.validateToken(request.token)
        return ResponseEntity.ok(TokenResponse(claims.subject, true, claims.get("role", String::class.java)))
    }
}
